File: src/firefly/java/edu/caltech/ipac/visualize/plot/RangeValues.java
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */
package edu.caltech.ipac.visualize.plot;


import java.io.Serializable;

/**
 * LZ 6/11/15
 * Modified in order to run the new stretch algorithm STRETCH_ASINH and power law gamma
 *
 * 6/26/15
 * Add ZP and WP codes
 */

public final class RangeValues implements Cloneable, Serializable {

    public static final int PERCENTAGE = 88;
    //public static final int MAXMIN     = 89;  // obsolete
    public static final int ABSOLUTE   = 90;
    public static final int ZSCALE     = 91;
    public static final int SIGMA      = 92;

    private static final double ASINH_Q =  Double.NaN; // if NaN, Q will be estimated based on range;
    private static final double GAMMA=2.0;

    // used for hue-preserving rgb only
    private static final double ASINH_STRETCH_PARAM = Double.NaN; // if NaN, use zscale
    private static final double SCALING_K = 1;

    public static final int STRETCH_LINEAR= 44;
    public static final int STRETCH_LOG   = 45;
    public static final int STRETCH_LOGLOG= 46;
    public static final int STRETCH_EQUAL = 47;
    public static final int STRETCH_SQUARED = 48;
    public static final int STRETCH_SQRT    = 49;
    public static final int STRETCH_ASINH   = 50;
    public static final int STRETCH_POWERLAW_GAMMA   = 51;

    private static final short RGB_PRESERVE_HUE_DEFAULT = 0; // use 0 for false, 1 for true


    private final int    _lowerWhich;
    private final double _lowerValue;
    private final int    _upperWhich;
    private final double _upperValue;
    private double _asinhQValue;
    private final double _gammaValue;
    private final int    _algorithm;
    private final short  _rgbPreserveHue;
    private double _asinhStretch; // used with hue-preserving rgb only
    private final double _scalingK; /* scaling factor for flux, used with hue-preserving rgb only */
    private final int    _zscale_contrast;
    private final int    _zscale_samples; /* desired number of pixels in sample */
    private final int    _zscale_samples_per_line; /* optimal number of pixels per line */
    private final double _bias;
    private final double _contrast;

    public RangeValues() {
        this( PERCENTAGE, 1.0, PERCENTAGE, 99.0, ASINH_Q, GAMMA, STRETCH_LINEAR, 25, 600, 120, RGB_PRESERVE_HUE_DEFAULT, ASINH_STRETCH_PARAM, SCALING_K);
    }

    public RangeValues( int    lowerWhich,
                        double lowerValue,
                        int    upperWhich,
                        double upperValue,
                        double asinhQValue,
                        double gammaValue,
                        int    algorithm,
                        int    zscale_contrast,
                        int    zscale_samples,
                        int    zscale_samples_per_line,
                        short  rgbPreserveHue,
                        double asinhStretch,
                        double scalingK) {
        this( lowerWhich, lowerValue, upperWhich, upperValue, asinhQValue, gammaValue, algorithm,
                zscale_contrast, zscale_samples, zscale_samples_per_line, rgbPreserveHue, asinhStretch, scalingK,
                0.5, 1.0);
    }


    public RangeValues( int    lowerWhich,
                        double lowerValue,
                        int    upperWhich,
                        double upperValue,
                        double asinhQValue,
                        double gammaValue,
                        int    algorithm,
                        int    zscale_contrast,
                        int    zscale_samples,
                        int    zscale_samples_per_line,
                        short  rgbPreserveHue,
                        double asinhStretch,
                        double scalingK,
                        double bias,
                        double contrast) {

        _lowerWhich= lowerWhich;
        _lowerValue= lowerValue;
        _upperWhich= upperWhich;
        _upperValue= upperValue;
        _algorithm = (rgbPreserveHue > 0) ? STRETCH_ASINH : algorithm;
        _asinhQValue = asinhQValue;
        _gammaValue=gammaValue;
        _zscale_contrast = zscale_contrast;
        _zscale_samples = zscale_samples;
        _zscale_samples_per_line = zscale_samples_per_line;
        _rgbPreserveHue = rgbPreserveHue;
        _asinhStretch = asinhStretch;
        _scalingK = scalingK;
        _bias = bias;
        _contrast = contrast;
    }

    public double getAsinhQValue() { return _asinhQValue; }
    public void setAsinhQValue(double val) { _asinhQValue = val; }
    public double getGammaValue() { return _gammaValue; }
    public double getAsinhStretch() { return _asinhStretch; }
    public void setAsinhStretch(double val) { _asinhStretch = val; }
    public double getScalingK() { return _scalingK; }

    public int    getLowerWhich() { return _lowerWhich; }
    public double getLowerValue() { return _lowerValue; }

    public int    getUpperWhich() { return _upperWhich; }
    public double getUpperValue() { return _upperValue; }

    public int getZscaleContrast() { return _zscale_contrast; }
    public int getZscaleSamples() { return _zscale_samples; }
    public int getZscaleSamplesPerLine() { return _zscale_samples_per_line; }

    public int getStretchAlgorithm() { return _algorithm; }

    public boolean rgbPreserveHue() { return _rgbPreserveHue != 0; }

    public Object clone() {
        return new RangeValues( _lowerWhich, _lowerValue, _upperWhich,
		_upperValue, _asinhQValue, _gammaValue, _algorithm,
		_zscale_contrast, _zscale_samples, _zscale_samples_per_line, _rgbPreserveHue, _asinhStretch, _scalingK, _bias, _contrast );
    }

    public String toString() { return serialize(); }

    public static RangeValues parse(String sIn) {
        if (isEmpty(sIn)) return null;

        try {
            String[] s= sIn.split(",");
            int i= 0;
            int    lowerWhich=              Integer.parseInt(s[i++]);
            double lowerValue=              parseDouble(s[i++]);
            int    upperWhich=              Integer.parseInt(s[i++]);
            double upperValue=              parseDouble(s[i++]);
            double asinhQValue=             parseDouble(s[i++]);
            double gammaValue=              parseDouble(s[i++]);
            int    algorithm=               Integer.parseInt(s[i++]);
            int    zscale_contrast=         Integer.parseInt(s[i++]);
            int    zscale_samples=          Integer.parseInt(s[i++]);
            int    zscale_samples_per_line= Integer.parseInt(s[i++]);
            short  rgbPreserveHue=          s.length > 10 ? Short.parseShort(s[i++]) : RGB_PRESERVE_HUE_DEFAULT;
            double asinhStretch=            s.length > 10 ? parseDouble(s[i++]) : ASINH_STRETCH_PARAM;
            double scalingK=                s.length > 10 ? parseDouble(s[i]) : SCALING_K;

            return new RangeValues(lowerWhich,
                    lowerValue,
                    upperWhich,
                    upperValue,
                    asinhQValue,
                    gammaValue,
                    algorithm,
                    zscale_contrast,
                    zscale_samples,
                    zscale_samples_per_line,
                    rgbPreserveHue,
                    asinhStretch,
                    scalingK);

        } catch (Exception e) {
            return null;
        }

    }


    public String serialize() {

        return getLowerWhich()+","+
                getLowerValue()+","+
                getUpperWhich()+","+
                getUpperValue()+","+
                getAsinhQValue()+","+
                getGammaValue()+","+
                getStretchAlgorithm()+","+
                getZscaleContrast()+","+
                getZscaleSamples()+","+
                getZscaleSamplesPerLine()+","+
                _rgbPreserveHue+","+
                getAsinhStretch()+","+
                getScalingK();
    }

    private static boolean isEmpty(String s) { return s == null || s.trim().length() == 0; }

    public static double parseDouble(String s) throws NumberFormatException {
        return s.equals("NaN") ? Double.NaN : Double.parseDouble(s);
    }
}

File: src/firefly/js/visualize/RangeValues.js
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */

import {isNumber, isString, isObject} from 'lodash';


export const PERCENTAGE_STR = 'percent';
export const ABSOLUTE_STR   = 'absolute';
export const SIGMA_STR      = 'sigma';

export const PERCENTAGE = 88;
// export const MINMAX     = 89;  // obsolete
export const ABSOLUTE   = 90;
export const ZSCALE     = 91;
export const SIGMA      = 92;

// the values should match the constants in RangeValues.java
export const STRETCH_LINEAR= 44;
export const STRETCH_LOG   = 45;
export const STRETCH_LOGLOG= 46;
export const STRETCH_EQUAL = 47;
export const STRETCH_SQUARED = 48;
export const STRETCH_SQRT    = 49;
export const STRETCH_ASINH   = 50;
export const STRETCH_POWERLAW_GAMMA   = 51;

const LINEAR_STR= 'linear';

const alStrToConst = {
    log : STRETCH_LOG,
    loglog : STRETCH_LOGLOG,
    equal : STRETCH_EQUAL,
    squared : STRETCH_SQUARED,
    sqrt : STRETCH_SQRT,
    asinh : STRETCH_ASINH,
    powerlaw_gamma : STRETCH_POWERLAW_GAMMA,
};


const boundsStrToConst = {
    percent:  PERCENTAGE,
    absolute: ABSOLUTE,
    sigma: SIGMA
};

const RGB_PRESERVE_HUE_DEFAULT= 0; // user 0 for false

export class RangeValues {
    constructor( lowerWhich= PERCENTAGE,
                 lowerValue= 1.0,
                 upperWhich= PERCENTAGE,
                 upperValue= 99.0,
                 asinhQValue=Number.NaN,
                 gammaValue=2.0,
                 algorithm= STRETCH_LINEAR,
                 zscaleContrast= 25,
                 zscaleSamples= 600,
                 zscaleSamplesPerLine= 120,
                 rgbPreserveHue= RGB_PRESERVE_HUE_DEFAULT,
                 asinhStretch=Number.NaN,
                 scalingK=1.0,
                 bias= 0.5,
                 contrast= 1.0 ) {
        this.lowerWhich= parseInt(lowerWhich+'');
        this.lowerValue= parseFloat(lowerValue+'');
        this.upperWhich= parseInt(upperWhich+'');
        this.upperValue= parseFloat(upperValue+'');
        this.asinhQValue = parseFloat(asinhQValue+'');
        this.gammaValue=parseFloat(gammaValue+'');
        this.algorithm=  parseInt(algorithm+'');
        this.zscaleContrast= parseInt(zscaleContrast+'');
        this.zscaleSamples= parseInt(zscaleSamples+''); /* desired number of pixels in sample */
        this.zscaleSamplesPerLine= parseInt(zscaleSamplesPerLine+''); /* optimal number of pixels per line */
        this.rgbPreserveHue=  parseInt(rgbPreserveHue+''); /* if 0, stretch by band, otherwise preserve hue*/
        this.asinhStretch=parseFloat(asinhStretch+'');
        this.scalingK=parseFloat(scalingK+'');
        if (this.rgbPreserveHue > 0) {
            this.algorithm = STRETCH_ASINH;
            this.upperWhich = ZSCALE;
        }
        this.bias= parseFloat(bias+'');
        this.contrast= parseFloat(contrast+'');
    }



    /**
     * @return {RangeValues}
     */
    clone() {
        return new RangeValues( this.lowerWhich, this.lowerValue, this.upperWhich,
            this.upperValue, this.asinhQValue,  this.gammaValue, this.algorithm,
            this.zscaleContrast, this.zscaleSamples, this.zscaleSamplesPerLine,
            this.rgbPreserveHue,this.asinhStretch,this.scalingK,
            this.bias, this.contrast );
    }

    toString() { return this.toJSON(); }


    /**
     *
     * @param p the params objects
     * @param p.which
     * @param p.lowerWhich
     * @param p.lowerValue
     * @param p.upperWhich
     * @param p.upperValue
     * @param p.asinhQValue
     * @param p.gammaValue
     * @param p.algorithm
     * @param p.zscaleContrast
     * @param p.zscaleSamples
     * @param p.zscaleSamplesPerLine
     * @param p.rgbPreserveHue
     * @param p.asinhStretch - stretch parameter for hue-preserving rgb
     * @param p.scalingK - flux scaling coefficient for hue-preserving rgb
     * @param p.bias
     * @param p.contrast
     * @return {RangeValues}
     */
    static makeRV({which= PERCENTAGE,
                      lowerWhich,
                      lowerValue= 1.0,
                      upperWhich,
                      upperValue= 99.0,
                      asinhQValue= Number.NaN,
                      gammaValue=2.0,
                      algorithm= STRETCH_LINEAR,
                      zscaleContrast= 25,
                      zscaleSamples= 600,
                      zscaleSamplesPerLine= 120,
                      rgbPreserveHue= RGB_PRESERVE_HUE_DEFAULT,
                      asinhStretch= Number.NaN,
                      scalingK=1.0,
                      bias= 0.5,
                      contrast= 1.0} ) {

        lowerWhich= lowerWhich || which;
        upperWhich= upperWhich || which;
        return new RangeValues( lowerWhich, lowerValue, upperWhich, upperValue, asinhQValue,
            gammaValue, algorithm, zscaleContrast, zscaleSamples,
            zscaleSamplesPerLine, rgbPreserveHue, asinhStretch, scalingK, bias, contrast);
    }

    /**
     * 
     * @param lowerWhich
     * @param lowerValue
     * @param upperWhich
     * @param upperValue
     * @param asinhQValue
     * @param gammaValue
     * @param algorithm
     * @param zscaleContrast
     * @param zscaleSamples
     * @param zscaleSamplesPerLine
     * @param rgbPreserveHue
     * @param asinhStretch used for hue preserving rgb, when NaN use zscale to estimate stretch value
     * @param scalingK used for hue preserving rgb
     * @param bias
     * @param contrast
     * @return {RangeValues}
     */
    static make(lowerWhich= PERCENTAGE,
                lowerValue= 1.0,
                upperWhich= PERCENTAGE,
                upperValue= 99.0,
                asinhQValue= Number.NaN,
                gammaValue= 2.0,
                algorithm= STRETCH_LINEAR,
                zscaleContrast= 25,
                zscaleSamples= 600,
                zscaleSamplesPerLine= 120,
                rgbPreserveHue= RGB_PRESERVE_HUE_DEFAULT,
                asinhStretch= Number.NaN,
                scalingK=1.0,
                bias= 0.5,
                contrast= 1.0 ) {
        return new RangeValues( lowerWhich, lowerValue, upperWhich, upperValue, asinhQValue,
             gammaValue, algorithm, zscaleContrast, zscaleSamples,
            zscaleSamplesPerLine, rgbPreserveHue, asinhStretch, scalingK, bias, contrast);
    }
    
    /**
     *
     * @param boundsType one of 'percent', 'absolute', 'sigma'
     * @param lowerValue lower value of stretch, based on stretchType
     * @param upperValue upper value of stretch, based on stretchType
     * @param algorithm one of 'linear', 'log', 'loglog', 'equal', 'squared', 'sqrt', 'asinh', powerlaw_gamma'
     * @return {*}
     */
    static makeSimple(boundsType= PERCENTAGE_STR,
                lowerValue= 1.0,
                upperValue= 1.0,
                algorithm= LINEAR_STR) {

        let btValue= PERCENTAGE;
        if (boundsStrToConst[boundsType.toLowerCase()]) {
            btValue= boundsStrToConst[boundsType.toLowerCase()];
        }

        let a= STRETCH_LINEAR;
        if (alStrToConst[algorithm.toLowerCase()]) {
            a= alStrToConst[algorithm.toLowerCase()];
        }

        return RangeValues.make( btValue, lowerValue, btValue, upperValue,Number.NaN,2, a);

    }

    static make2To10SigmaLinear() {
        return RangeValues.makeRV({which:SIGMA, lowerValue:-2, upperValue:10, algorithm:STRETCH_LINEAR});
    }

    /**
     *
     * @param boundsType one of 'percent', 'absolute', 'sigma'
     * @param lowerValue lower value of stretch, based on stretchType
     * @param upperValue upper value of stretch, based on stretchType
     * @param algorithm one of 'log', 'loglog', 'equal', 'squared', 'sqrt', 'asinh', powerlaw_gamma'
     * @return {*}
     */
    static serializeSimple(boundsType= PERCENTAGE_STR,
                           lowerValue= 1.0,
                           upperValue= 1.0,
                           algorithm= LINEAR_STR ) {
        const rv= RangeValues.makeSimple(boundsType,lowerValue,upperValue,algorithm);
        return rv.toJSON();
    }

    /**
     * @param sIn serialized string representation of RangeValues
     * @return {RangeValues}
     */
    static parse(sIn) {
        if (!sIn) return undefined;
        if (isObject(sIn) &&
            isNumber(sIn.lowerWhich) && isNumber(sIn.upperWhich) &&
            isNumber(sIn.lowerValue) && isNumber(sIn.upperValue) && isNumber(sIn.algorithm)) {
            return sIn;
        }
        if (!isString(sIn)) return undefined;


        const params= sIn.split(',').map( (v) => Number.parseFloat(v) );
        const valid= params.every( (v)=> typeof v !== 'undefined');

        return valid ? new RangeValues(...params) : undefined;
    }

    toJSON() {
        return RangeValues.serializeRV(this);
    }

    static serializeRV(rv) {
        return rv.lowerWhich+','+
            rv.lowerValue+','+
            rv.upperWhich+','+
            rv.upperValue+','+
            rv.asinhQValue+','+
            rv.gammaValue+','+
            rv.algorithm+','+
            rv.zscaleContrast+','+
            rv.zscaleSamples+','+
            rv.zscaleSamplesPerLine+','+
            rv.rgbPreserveHue+','+
            rv.asinhStretch+','+
            rv.scalingK;
    }
    
}

export default RangeValues;
